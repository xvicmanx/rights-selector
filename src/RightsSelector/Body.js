import React from "react";
import PropTypes from 'prop-types';

const Body = props => (
  <tbody>
    {props.resources.map(resource => (
      <tr key={`${resource.value}`}>
        <th>{resource.label}</th>
        {props.actions.map(action => (
          <td key={`${action.value}::${resource.value}`}>
            <input
              onClick={props.onClick}
              type="checkbox"
              value={`${action.value}::${resource.value}`}
            />
          </td>
        ))}
      </tr>
    ))}
  </tbody>
);

Body.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.any,
    })
  ).isRequired,
  resources: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.any,
    })
  ).isRequired,
  onClick: PropTypes.func.isRequired,
};


export default Body;
