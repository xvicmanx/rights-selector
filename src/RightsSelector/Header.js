import React from "react";
import PropTypes from 'prop-types';

const Header = props => (
  <thead>
    <tr>
      <th />
      <th colSpan={props.actions.length}>{props.actionsLabel}</th>
    </tr>
    <tr>
      <th>{props.resourcesLabel}</th>
      {props.actions.map(action => <th key={action.value}>{action.label}</th>)}
    </tr>
  </thead>
);

Header.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
    })
  ).isRequired,
  actionsLabel: PropTypes.string.isRequired,
  resourcesLabel: PropTypes.string.isRequired,
};


export default Header;
