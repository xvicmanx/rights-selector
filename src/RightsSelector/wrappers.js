import bcc from 'bem-react-component-creator';

const { block, element } = bcc('rights-selector');

const Container = block('table');
Container.Header = element('thead', 'header');
Container.Body = element('tbody', 'body');
Container.Row = element('tr', 'row');
Container.HeaderCell = element('th', 'header-cell');
Container.Cell = element('td', 'cell');
Container.Input = element('input', 'input');

export default Container;