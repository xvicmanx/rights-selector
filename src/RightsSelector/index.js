import React from 'react';
import PropTypes from 'prop-types';
import Header from './Header';
import Body from './Body';

const RIGHT_REGEX = /(.+)::(.+)/;

class RightsSelector extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      rights: props.rights || {}
    };
    this.handleClick = this.handleClick.bind(this);
  }

  handleClick(evt) {
    const { value } = evt.target;
    const { rights } = this.state;
    const action = value.replace(RIGHT_REGEX, "$1");
    const resource = value.replace(RIGHT_REGEX, "$2");
    const update = { ...rights };
    update[resource] = update[resource] || {};
    update[resource][action] = !update[resource][action];
    this.setState({ rights: update });
    this.props.onChange(update);
  }

  render() {
    const {
      actions,
      resources,
      actionsLabel,
      resourcesLabel,
    } = this.props;
    return (
      <table>
        <Header
          actions={actions}
          resourcesLabel={resourcesLabel}
          actionsLabel={actionsLabel}
        />
        <Body
          actions={actions}
          resources={resources}
          onClick={this.handleClick}
        />
      </table>
    );
  }
}

RightsSelector.propTypes = {
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.any,
    })
  ).isRequired,
  resources: PropTypes.arrayOf(
    PropTypes.shape({
      label: PropTypes.string,
      value: PropTypes.any,
    })
  ).isRequired,
  onChange: PropTypes.func.isRequired,
  actionsLabel: PropTypes.string.isRequired,
  resourcesLabel: PropTypes.string.isRequired,
};

export default RightsSelector;
