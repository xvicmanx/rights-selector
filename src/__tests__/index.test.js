import React from 'react';
import RightsSelector from '../RightsSelector/index';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import Enzyme, { shallow, mount, render } from 'enzyme';

Enzyme.configure({ adapter: new Adapter() })

describe('RightsSelector Component', () => {

  it('renders correctly', () => {
    const tree = renderer
      .create(<RightsSelector
        actionsLabel="Actions"
        resourcesLabel="Resources"
        actions={[
          {
            value: 'create',
            label: 'Create',
          },
          {
            value: 'read',
            label: 'Read',
          },
          {
            value: 'update',
            label: 'Update',
          },
          {
            value: 'delete',
            label: 'Delete',
          },
        ]}
        resources={[
          {
            value: 'book',
            label: 'Book',
          },
          {
            value: 'author',
            label: 'Author',
          }
        ]}
        onChange={(values) => {
          console.log(values);
        }}
      />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
})

