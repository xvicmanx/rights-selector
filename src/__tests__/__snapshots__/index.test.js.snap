// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RightsSelector Component renders correctly 1`] = `
<table>
  <thead>
    <tr>
      <th />
      <th
        colSpan={4}
      >
        Actions
      </th>
    </tr>
    <tr>
      <th>
        Resources
      </th>
      <th>
        Create
      </th>
      <th>
        Read
      </th>
      <th>
        Update
      </th>
      <th>
        Delete
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>
        Book
      </th>
      <td>
        <input
          onClick={[Function]}
          type="checkbox"
          value="create::book"
        />
      </td>
      <td>
        <input
          onClick={[Function]}
          type="checkbox"
          value="read::book"
        />
      </td>
      <td>
        <input
          onClick={[Function]}
          type="checkbox"
          value="update::book"
        />
      </td>
      <td>
        <input
          onClick={[Function]}
          type="checkbox"
          value="delete::book"
        />
      </td>
    </tr>
    <tr>
      <th>
        Author
      </th>
      <td>
        <input
          onClick={[Function]}
          type="checkbox"
          value="create::author"
        />
      </td>
      <td>
        <input
          onClick={[Function]}
          type="checkbox"
          value="read::author"
        />
      </td>
      <td>
        <input
          onClick={[Function]}
          type="checkbox"
          value="update::author"
        />
      </td>
      <td>
        <input
          onClick={[Function]}
          type="checkbox"
          value="delete::author"
        />
      </td>
    </tr>
  </tbody>
</table>
`;
